### PR TITLE
DnsNameResolver: Add DnsQueryIdSpace class to reduce overhead while g…

### PR DESCRIPTION
**Pull request by normanmaurer**
…enerating IDs

Motivation:

How we currently implement the selection of the next id to use for a query is very inefficient, especially once we already have a lot of inflight queries. This is problematic as we basically "block" the eventloop while doing the for loop.

Modifications:

Add new datastructure for storing and selecting ids that is basically O(1)

Result:

Less overhead while selecting ids

### CODE REVIEW

**Comment by normanmaurer** [comment]
I am exploring some other idea's as well as this basically will need 131k per instance, (which might be ok as its per nameserver that is configured for a `DnsNameResolver` and I don't expect people to have more then 10 name servers configured at max.)

File: resolver-dns/src/main/java/io/netty/resolver/dns/DnsQueryIdSpace.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                idx = random.nextInt(tail - head) + head;
+            }
+            short otherId = ids[idx];
+            ids[idx] = (short) id;
+            ids[tail] = otherId;
+            assert otherId != (short) id;
+        }
+        tail = (tail + 1) & mask;
+        count++;
+    }
```
Comment (bryce-anderson) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsQueryIdSpace.java:95]: I think you can make this substantially simpler by just making the insertion position random and always pulling from the tail. That leaves the only state left being `ids` and `count`.
```suggestion
    void pushId(int id) {
        if (count == ids.length) {
            throw new IllegalStateException("overflow");
        }

        assert id <= MAX_ID && id >= 0;
        // pick a slot for our index, and whatever was in that slot before will get moved to the tail.
        Random random = PlatformDependent.threadLocalRandom();
        int insertionPosition = random.nextInt(count + 1);
        ids[count + 1] = ids[insertionPosition];
        ids[insertionPosition] = (short) id;
        count++;
    }
```

Comment (bryce-anderson) [reply resolver-dns/src/main/java/io/netty/resolver/dns/DnsQueryIdSpace.java:95]: And if you want to make it more obviously random we can move this to the `nextId()` method where it simply selects a random position up to `count`, used that as the result, and moves the tail into the hole that was made.

File: resolver-dns/src/main/java/io/netty/resolver/dns/DnsQueryIdSpace.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+/**
+ * Special data-structure that will allow to retrieve the next query id to use, while still guarantee some sort
+ * of randomness.
+ * The query id will be between 0 (inclusive) and 65535 (inclusive) as defined by the RFC.
+ */
+final class DnsQueryIdSpace {
+    private static final int MAX_ID = 65535;
+
+    // Holds all possible ids which are stored as unsigned shorts
+    private final short[] ids = new short[MAX_ID + 1];
```
Comment (bryce-anderson) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsQueryIdSpace.java:32]: Perhaps for a future optimization we could make this growable and start small. That should keep it roughly on the same footprint as the `Map` that holds the state objects.

Comment (bryce-anderson) [reply resolver-dns/src/main/java/io/netty/resolver/dns/DnsQueryIdSpace.java:32]: Scratch this, sorry: it's actually the opposite: the size of this is at it's max when no id's have been issues which is the exact opposite of the `Map` that holds the state objects.

Comment (normanmaurer) [reply resolver-dns/src/main/java/io/netty/resolver/dns/DnsQueryIdSpace.java:32]: Yeah... I was also thinking about ways how to grow / shrink but all I came up with would result in less randomness which I think is not good

File: resolver-dns/src/main/java/io/netty/resolver/dns/DnsQueryIdSpace.java (hunk 1 above)
Comment (normanmaurer) [reply resolver-dns/src/main/java/io/netty/resolver/dns/DnsQueryIdSpace.java]: Thanks a lot! This is a great idea <3 . I adjusted the code to follow your idea

**Review by bryce-anderson — APPROVED** [review]
lgtm. I know you may still iterate but if you decide to try this for now it looks good.

**Comment by normanmaurer** [comment]
Closing this... we will use https://github.com/netty/netty/pull/13795/

**Pull request closed without merging**